src/quickgui/qgsquickpositionkit.cpp

```diff
@@ -0,0 +1,311 @@
+    mSource->startUpdates();
+
+    QgsDebugMsg( QStringLiteral( "Position source changed: %1" ).arg( mSource->sourceName() ) );
+  }
+}
+
+QgsQuickMapSettings *QgsQuickPositionKit::mapSettings() const
+{
+  return mMapSettings;
+}
+
+void QgsQuickPositionKit::updateProjectedPosition()
+{
+  if ( !mMapSettings )
+    return;
+
+  QgsPointXY srcPoint = QgsPointXY( mPosition.x(), mPosition.y() );
+  QgsPointXY projectedPositionXY = QgsQuickUtils::transformPoint( QgsQuickUtils::coordinateReferenceSystemFromEpsgId( 4326 ), mMapSettings->destinationCrs(), mMapSettings->transformContext(), srcPoint );
+  mProjectedPosition = QgsPoint( projectedPositionXY.x(), projectedPositionXY.y() );
+  mProjectedPosition.addZValue( mPosition.z() );
+}
+
+void QgsQuickPositionKit::positionUpdated( const QGeoPositionInfo &info )
+{
+  if ( !info.coordinate().isValid() )
+  {
+    // keep last valid position
+    mHasPosition = false;
+    emit hasPositionChanged();
+  }
+
+
+  mPosition = QgsPoint( info.coordinate().longitude(),
+                        info.coordinate().latitude(),
+                        info.coordinate().altitude() ); // can be NaN
+
+  if ( info.hasAttribute( QGeoPositionInfo::HorizontalAccuracy ) )
+    mAccuracy = info.attribute( QGeoPositionInfo::HorizontalAccuracy );
+  else
+    mAccuracy = -1;
+  mScreenAccuracy = calculateScreenAccuracy();
+
+  if ( info.hasAttribute( QGeoPositionInfo::Direction ) )
+    mDirection = info.attribute( QGeoPositionInfo::Direction );
+  else
+    mDirection = -1;
+
+  emit positionChanged();
+
+  if ( !mHasPosition )
+  {
+    mHasPosition = true;
+    emit hasPositionChanged();
+  }
+}
+void QgsQuickPositionKit::onSimulatePositionLongLatRadChanged( QVector<double> simulatePositionLongLatRad )
+{
+  if ( simulatePositionLongLatRad.size() > 2 )
+  {
+    double longitude = simulatePositionLongLatRad[0];
+    double latitude = simulatePositionLongLatRad[1];
+    double radius = simulatePositionLongLatRad[2];
+    QgsDebugMsg( QStringLiteral( "Use simulated position around longlat: %1, %2, %3" ).arg( longitude ).arg( latitude ).arg( radius ) );
+    use_simulated_location( longitude, latitude, radius );
+  }
+  else
+  {
+    QgsDebugMsg( QStringLiteral( "Unable to set simulated position due to the input errors." ) );
+    use_gps_location();
+  }
+}
+
+double QgsQuickPositionKit::calculateScreenAccuracy()
+{
+  if ( !mMapSettings )
+    return 2.0;
+
+  if ( accuracy() > 0 )
+  {
+    double scpm = QgsQuickUtils::screenUnitsToMeters( mMapSettings, 1 );
+    if ( scpm > 0 )
+      return 2 * ( accuracy() / scpm );
+    else
+      return 2.0;
+  }
+  return 2.0;
+}
+
+void QgsQuickPositionKit::onUpdateTimeout()
+{
+  if ( mHasPosition )
+  {
+    mHasPosition = false;
+    emit hasPositionChanged();
+  }
+}
+
+QPointF QgsQuickPositionKit::screenPosition() const
+{
+  return mScreenPosition;
+}
+
+void QgsQuickPositionKit::setScreenPosition( const QPointF &screenPosition )
+{
+  mScreenPosition = screenPosition;
```

Comment:
Is the notification signal not sent here on purpose? How does this method relate to updateScreenPosition()?
Hmm... actually, is the screen position not a derived (readonly) property?